File: template/sbtree/sbtree_test.go
package sbtree_test

import (
	"math/rand"
	"testing"
	"time"

	"github.com/spinlock/go-libs/assert"
	. "github.com/spinlock/snippets/template/sbtree"
)

func checkTree(t *Tree, expect map[int]bool) {
	assert.Must(len(expect) == t.Size())
	assert.Must(t.Check())

	nodes := t.Nodes(nil)
	assert.Must(len(expect) == t.Size())
	for i := 0; i < len(nodes); i++ {
		assert.Must(t.Rank(nodes[i].Key()) == i)
		x := t.Select(i)
		assert.Must(x != nil && x.Key() == nodes[i].Key())
	}

	for i := 1; i < len(nodes); i++ {
		x := t.FindPred(nodes[i].Key())
		assert.Must(x != nil && x.Key() == nodes[i-1].Key())
	}
	for i := 0; i < len(nodes)-1; i++ {
		x := t.FindSucc(nodes[i].Key())
		assert.Must(x != nil && x.Key() == nodes[i+1].Key())
	}

	for k, _ := range expect {
		assert.Must(t.Contains(k))
	}

	for i := 1; i < len(nodes); i++ {
		assert.Must(nodes[i].Key() > nodes[i-1].Key())
	}
	for i := 0; i < len(nodes); i++ {
		assert.Must(expect[nodes[i].Key()])
	}
}

func TestInsert(x *testing.T) {
	t := New()
	checkTree(t, map[int]bool{})

	m := make(map[int]bool)
	for i := 0; i < 1024; i++ {
		m[i] = true
		_, addNode := t.Insert(i, nil)
		assert.Must(addNode)
	}
	checkTree(t, m)

	for i := 0; i < 1024; i++ {
		_, addNode := t.Insert(i, nil)
		assert.Must(addNode == false)
	}
	checkTree(t, m)
}

func TestRemove(x *testing.T) {
	t := New()
	for i := 0; i < 1024; i++ {
		_, addNode := t.Insert(i, i)
		assert.Must(addNode)
	}

	for i := 0; i < 1024; i++ {
		oldValue, delNode := t.Remove(i)
		assert.Must(delNode)
		assert.Must(oldValue.(int) == i)
	}
	checkTree(t, map[int]bool{})

	for i := 0; i < 1024; i++ {
		oldValue, delNode := t.Remove(i)
		assert.Must(delNode == false)
		assert.Must(oldValue == nil)
	}
	checkTree(t, map[int]bool{})
}

func TestRandom(x *testing.T) {
	t := New()
	m := make(map[int]bool)
	r := rand.New(rand.NewSource(time.Now().UnixNano()))

	for i := 0; i < 4096; i++ {
		x := int(r.Int31n(65536))
		m[x] = true
		t.Insert(x, nil)
	}
	checkTree(t, m)

	for k, _ := range m {
		delete(m, k)
		oldValue, delNode := t.Remove(k)
		assert.Must(delNode)
		assert.Must(oldValue == nil)
	}
	checkTree(t, m)
}

File: src/test/java/template/SBTreeTest.java
package template;

import org.junit.Test;

import java.util.ArrayList;

import static org.junit.Assert.assertEquals;

public class SBTreeTest {

    @Test
    public void testAdd() {
        SBTree<Integer> tree = new SBTree<>();
        for (int i = 0; i < 1024; i++) {
            tree.add(i);
            assertEquals(i + 1, tree.size());
        }
        for (int i = 0; i < 1024; i++) {
            tree.add(i);
            assertEquals(1024, tree.size());
        }

        ArrayList<Integer> values = new ArrayList<>(tree.toList());
        assertEquals(1024, values.size());
        for (int i = 0; i < values.size(); i++) {
            assertEquals(i, (int) values.get(i));
        }
    }

    @Test
    public void testRemove() {
        SBTree<Integer> tree = new SBTree<>();
        for (int i = 0; i < 1024; i++) {
            tree.add(i);
            assertEquals(i + 1, tree.size());
        }
        for (int i = 0; i < 1024; i++) {
            tree.remove(i);
            assertEquals(1023 - i, tree.size());
        }
    }

    @Test
    public void testRank() {
        SBTree<Integer> tree = new SBTree<>();
        for (int i = 0; i < 1024; i++) {
            tree.add(i);
        }
        for (int i = 0; i < 1024; i++) {
            assertEquals(i, tree.rank(i));
        }
        assertEquals(-1, tree.rank(-1));
        assertEquals(-1025, tree.rank(1024));
    }

    @Test
    public void testSelect() {
        SBTree<Integer> tree = new SBTree<>();
        for (int i = 0; i < 1024; i++) {
            tree.add(i);
        }
        for (int i = 0; i < 1024; i++) {
            assertEquals(i, (int) tree.select(i));
        }
    }

}